sparse/dok.py

```diff
@@ -0,0 +1,312 @@
+import six
+
+import numpy as np
+
+# Zip with Python 2/3 compat
+# Consumes less memory than Py2 zip
+from six.moves import zip, range
+
+from numbers import Integral
+from collections import Iterable
+
+from .slicing import normalize_index
+from .utils import _zero_of_dtype
+
+try:  # Windows compatibility
+    int = long
+except NameError:
+    pass
+
+
+class DOK:
+    """
+    A class for building sparse multidimensional arrays.
+
+    Attributes
+    ----------
+    dtype : numpy.dtype
+        The datatype of this array. Can be :code:`None` if no elements
+        have been set yet.
+    shape : tuple[int]
+        The shape of this array.
+    dict : dict
+        The keys of this dictionary contain all the indices and the values
+        contain the nonzero entries.
+
+    See Also
+    --------
+    COO : A read-only sparse array.
+
+    Examples
+    --------
+    You can create :obj:`DOK` objects from Numpy arrays.
+
+    >>> x = np.eye(5, dtype=np.uint8)
+    >>> x[2, 3] = 5
+    >>> s = DOK(x)
+    >>> s
+    <DOK: shape=(5, 5), dtype=uint8, nnz=6>
+
+    You can also create them from just shapes, and use slicing assignment.
+
+    >>> s2 = DOK((5, 5))
+    >>> s2[1:3, 1:3] = [[4, 5], [6, 7]]
+    >>> s2
+    <DOK: shape=(5, 5), dtype=int64, nnz=4>
+
+    You can convert :obj:`DOK` arrays to :obj:`COO` arrays, or :obj:`numpy.ndarray`
+    objects.
+
+    >>> from sparse import COO
+    >>> s3 = COO(s2)
+    >>> s3
+    <COO: shape=(5, 5), dtype=int64, nnz=4, sorted=False, duplicates=False>
+    >>> s2.todense()  # doctest: +NORMALIZE_WHITESPACE
+    array([[0, 0, 0, 0, 0],
+           [0, 4, 5, 0, 0],
+           [0, 6, 7, 0, 0],
+           [0, 0, 0, 0, 0],
+           [0, 0, 0, 0, 0]])
+    """
+    def __init__(self, shape, values=None, dtype=None):
+        """
```

Comment:
I recommend only including the docstring on the main class and removing it from the `__init__` method.  People tend to check for docstrings like `DOK?` or `help(DOK)` rather than `help(DOK.__init__)`.  Also, the two versions here will almost certainly grow out of sync.

Reply:
That's how we're doing it in `COO`. I tend to keep the class docstring somewhat general (like an overview) and the `__init__` docstring specific to constructing objects.

Reply:
I recommend merging the docstring in `__init__` up to the class-level docstring.  I find that people rarely look at the `__init__` docstring in practice.